ggml-cuda.cu

```diff
@@ -2405,11 +2405,100 @@ GGML_CALL static void ggml_backend_cuda_synchronize(ggml_backend_t backend) {
+
 GGML_CALL static enum ggml_status ggml_backend_cuda_graph_compute(ggml_backend_t backend, ggml_cgraph * cgraph) {
     ggml_backend_cuda_context * cuda_ctx = (ggml_backend_cuda_context *)backend->context;
 
     ggml_cuda_set_device(cuda_ctx->device);
 
+#ifdef USE_CUDA_GRAPH
+    // Objects required for CUDA Graph
+    static ggml_cudaGraph cudaGraph;
+    bool useCudaGraph = (cudaGraph.count>=7); //avoid CUDA graphs on first few steps due to incompatible initialisations.
+    char** updatedKernelArg[MAX_NODES_IN_CUDA_GRAPH];
+    bool cudaGraphUpdateRequired = false;
+    // pointer to CUDA cpy kernel, which is required to identify
+    // kernel parameters which need updated in the graph for each token
+    void* ggmlCudaCpyFn = nullptr;
+
+    if(cudaGraph.count==0){        
+        cudaDeviceProp prop;
+        int device;
+        CUDA_CHECK(cudaGetDevice(&device));
+        CUDA_CHECK(cudaGetDeviceProperties(&prop, device));
+        if (prop.major < 8){
+            cudaGraph.disableDueToGpuArch=true;
+        }
+    }
+
+    // Disable CUDA graphs in presence of env var or old GPU.
+    // Also disable for multi-gpu for now. TO DO investigate
+    if(disableCudaGraphs || cudaGraph.disableDueToGpuArch || ggml_backend_cuda_get_device_count() > 1){
+        useCudaGraph = false;
+    }
+
+    if(useCudaGraph) {
+
+        if(cudaGraph.instance == nullptr) cudaGraphUpdateRequired=true;
+
+        // Loop over nodes in GGML graph to obtain info needed for CUDA graph
+        int k=0;
+        for (int i = 0; i < cgraph->n_nodes; i++) {
+            ggml_tensor * node = cgraph->nodes[i];
+            // Identify if the graph needs updated for this token due to the number of elements changing
+            // (identified by inspecting soft max op parameters)
+            if(node->op == GGML_OP_SOFT_MAX) {
+                if(node->src[1]->ne[1] > 1){
+                    useCudaGraph = false; // disable CUDA graphs for batch size > 1 for now. TO DO investigate
+                }
+                if(node->src[0]->ne[0] != cudaGraph.softmax_ne0) {
+                    cudaGraphUpdateRequired = true;
+                    cudaGraph.softmax_ne0 = node->src[0]->ne[0];
+                }
+            }
+            if(node->op == GGML_OP_CPY) {
+                // store the copy op parameter which changes with each token.
+                updatedKernelArg[k++]=(char**) &(node->src[1]->data);
+                if(ggmlCudaCpyFn == nullptr){
+                    // store a pointer to the copy op CUDA kernel to identify it later
+                    ggmlCudaCpyFn = ggml_cuda_cpy_fn(node->src[0], node->src[1]);
+                }
+            }
+        }
+    }
+
+    if(useCudaGraph && cudaGraphUpdateRequired) { // Start CUDA graph capture
+        CUDA_CHECK(cudaStreamBeginCapture(cuda_ctx->stream(), cudaStreamCaptureModeGlobal));
+    }
+
+#else
+    bool useCudaGraph = false;
+    bool cudaGraphUpdateRequired = false;
+#endif
```

Comment:
```suggestion
#endif // USE_CUDA_GRAPH
```

Reply:
Done (and similar at the end of the function)